### PR TITLE
Allow utilities to be loaded without RequireJS

### PR DESCRIPTION
**Pull request by andy-armstrong**
This commit adds the ability for utility classes to be loaded directly without RequireJS. If the class is loaded and RequireJS is not found, then the class will instead be installed into the edx namespace. For example, StringUtils can be accessed as ``edx.StringUtils``. 

The intention is that this will be useful for the LMS which is not fully RequireJS enabled. In particular, XBlocks and XModules have no ability to use RequireJS, so need global access to these utilities. This capability is being added to edx-platform in this PR: https://github.com/edx/edx-platform/pull/11904

In addition, HtmlUtils.template has been extended to pass StringUtils as a context parameter in addition to HtmlUtils itself.

@AlasdairSwan @cahrens please review.

FYI @bjacobel @dan-f @tobz @robrap @efischer19

### CODE REVIEW

**Comment by edx-webhook** [comment]
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

**Comment by andy-armstrong** [comment]
@AlasdairSwan I've been unable to debug why the Travis CI build keeps failing. Maybe we can look at this together on Monday,

File: src/js/utils/constants.js [hunk 1]
```diff
@@ -1,19 +1,25 @@
...
+                left: 37,
+                up: 38,
+                right: 39,
+                down: 40
+            }
+        };
+    });
+}).call(
+    this,
+    (typeof define !== 'undefined' && define) ||
```
Comment (cahrens) [inline src/js/utils/constants.js]: You shouldn't need the RequireJS check-- I was imagining you would use the exact syntax expected by the optimizer.

We use the namespace property, and the optimizer will rename things accordingly (assuming it is the expected syntax).

http://requirejs.org/docs/faq-advanced.html#rename

File: src/js/utils/global-loader.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+(function() {
+    'use strict';
+
+    window.edx = window.edx || {};
+
+    window.edx.GlobalLoader = (function() {
+        var registeredModules = {};
+
+        // Register standard libraries
+        registeredModules.jquery = $;
```
Comment (cahrens) [inline src/js/utils/global-loader.js]: Why is this necessary?

Comment (andy-armstrong) [inline src/js/utils/global-loader.js]: It's to make it so that HtmlUtils can load its dependencies:

https://github.com/edx/edx-ui-toolkit/blob/andya/minor-api-fixes/src/js/utils/html-utils.js#L9

File: src/js/utils/constants.js (hunk 1 above)
Comment (andy-armstrong) [inline src/js/utils/constants.js]: Thanks for the details about the syntax expected by the optimizer. I'll switch to that way of looking for ``define``. I don't understand why we don't need to use the ``RequireJS`` check though. This code is in the UI Toolkit and can be loaded in many different ways, including by code included in LMS factories where ``define`` is not defined and ``RequireJS.define`` is available. 

Let's discuss this in the morning and pin down right the combination of checks to be using in all these files. We might also want to recommend that LMS code use the same convention, rather than the current:

    define || RequireJS.define

I tried this originally here in the UI Toolkit but it doesn't work because ``define`` throws an undefined exception. The only reason it works in the LMS is due to this ugly hack where window is given a ``define`` property of undefined, rather than not having the property at all:

https://github.com/edx/edx-platform/blob/master/common/static/js/RequireJS-namespace-undefine.js#L7

For anyone reading along, here's an example of the standard idiom in the LMS:

https://github.com/edx/edx-platform/blob/master/lms/djangoapps/teams/static/teams/js/views/teams.js#L32

**Comment by andy-armstrong** [comment]
@cahrens In my testing it seems that the ``RequireJS`` namespace is still needed for the moment. We should continue the discussion as to how we might be able to remove it. I'm okay with including this in the UI Toolkit because a) this is an edX-specific repo, and b) it is written so that it won't fail if it isn't available.

I've also updated the non-utils to use the same mechanism to ensure that they will work in edx-platform, except I didn't add the logic to let them load outside of RequireJS. I really don't want to support the entire UI Toolkit outside RequireJS, so I think of the utilities as being exceptions.

File: src/js/dropdown-menu/dropdown-menu-view.js [hunk 3]
```diff
@@ -38,242 +39,247 @@ define([
...
-    }
+            });
+
+            return DropdownMenuView;
+        }
+    );
+}).call(
+    this,
+    (typeof define === 'function' && define.amd && define) ||
+    (typeof RequireJS !== 'undefined' && RequireJS.define)
```
Comment (AlasdairSwan) [inline src/js/dropdown-menu/dropdown-menu-view.js]: @andy-armstrong the diffs are showing weirdness for these files, is the only really difference the first few and last few lines of the file?

Comment (andy-armstrong) [inline src/js/dropdown-menu/dropdown-menu-view.js]: Try viewing the diffs with ``?w=1`` to see the non-whitespace changes:

https://github.com/edx/edx-ui-toolkit/pull/46/files?w=1

Comment (AlasdairSwan) [inline src/js/dropdown-menu/dropdown-menu-view.js]: Thanks for the tip, that's a lot less noisy

Comment (andy-armstrong) [inline src/js/dropdown-menu/dropdown-menu-view.js]: I don't know why GitHub keeps this feature so hidden. It is incredibly useful.

**Comment by AlasdairSwan** [comment]
Once the Travis failures are resolved :+1: Do you want to find some time tomorrow to debug (sorry today was a lot busier than I thought it would be)?

File: src/js/dropdown-menu/dropdown-menu-view.js [hunk 4]
```diff
@@ -38,242 +39,247 @@ define([
...
-        return DropdownMenuView;
-    }
+            });
+
+            return DropdownMenuView;
+        }
+    );
+}).call(
+    this,
+    (typeof define === 'function' && define.amd && define) ||
```
Comment (cahrens) [inline src/js/dropdown-menu/dropdown-menu-view.js]: I find this difficult to read (the last && define). I would prefer if clauses.

Comment (dan-f) [inline src/js/dropdown-menu/dropdown-menu-view.js]: Also, isn't that last `&& define` redundant with `typeof define === 'function'`?

Comment (andy-armstrong) [inline src/js/dropdown-menu/dropdown-menu-view.js]: The last ``&& define`` is what actually returns the value to be returned. It is a confusing syntax that only makes sense to ex-Lisp developers like me. :-) @cahrens and I have come up with a less concise but easier to comprehend version.

**Comment by andy-armstrong** [comment]
@cahrens @AlasdairSwan @dan-f I've addressed the review feedback so now all that remains is to a) get the Travis builds to pass, and b) to try it out on a sandbox.

**Comment by cahrens** [comment]
:+1: if it works. :)

**Comment by andy-armstrong** [comment]
@cahrens @AlasdairSwan strangely the Travis build just passed without me changing anything...

**Comment by AlasdairSwan** [comment]
Gotta love flaky Travis tests...
